iree/compiler/Conversion/CodegenUtils/ForOpCanonicalization.cpp

```diff
@@ -0,0 +1,155 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassRegistry.h"
+
+// Pass to combine instructions across ForOp boundary. It is common when doing
+// incremental lowering to generate transient ops that cancel each others out.
+// Canonicalization usually clean up those operations. When the value is loop
+// carried, MLIR canonicalization currently doesn't remove the redundant
+// operations.
+//
+// This pass allow to workaround MLIR limitation and does ad hoc clean up of
+// instructions found in IREE. Once we have a more general mechanism in MLIR
+// this pass can be completely removed.
+// This pass does this kind of transformation:
+// ```
+// %21 = vector.shape_cast %20 : vector<4xf32> to vector<1x4xf32>
+// %22 = scf.for %arg3 = %c0 to %c4096 step %c4 iter_args(%arg4 = %21)
+//    -> vector<1x4xf32> {
+//    [...]
+//    %100 = vector.shape_cast %arg4 : vector<1x4xf32> to vector<4xf32>
+//    [...]
+//    %109 = vector.shape_cast %108 : vector<4xf32> to vector<1x4xf32>
+//    scf.yield %109 : vector<1x4xf32>
+//  }
+//  %24 = vector.shape_cast %22 : vector<1x4xf32> to vector<4xf32>
+// ```
+// ->
+// ```
+// %22 = scf.for %arg3 = %c0 to %c4096 step %c4 iter_args(%arg4 = %20)
+//    -> vector<4xf32> {
+//    [...]
+//    scf.yield %108 : vector<4xf32>
+//  }
+// ```
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+class ForOpArgFolding final : public OpRewritePattern<scf::ForOp> {
```

Comment:
I am not sure such uber patterns work well. I would decompose this into multiple patterns.

- A pattern for `scf.for`. In this look at all uses of the arguments. If all uses are such that you can change the type of the argument in region and all so change the operand used in the for op. When you change the type of the argument you will have to do a `replaceAllUsesWith` to change all the uses
- Now the `shapeCastOp` and `ExtractOp` are in effectively a no-op. So separate patterns for the no-op.

Reply:
Just to clarify this a bit more (After thinking about this). Pattern mechanism doesnt really work here. We need to side-step it a bit. So the suggestion here is this

- Create a base class that derives from `OpRewritePattern` something like this:

```
template <typename DerivedTy>
class RegionOpFolderPattern : public OpRewritePattern<DerivedTy> {
   RegionOpFolderPattern(MLIRContext *context, ArrayRef<Value> regionOperands, PatternBenefit benefit = 1)
     OpRewritePattern(context, benefit),
     regionOperands(regionOperands)

protected : 
  ArrayRef<Value> regionOperands;
};

``` 

`regionOperands` is the list of operands of the surrounding region op, in this case the `ForOp`. 

Now add a pattern for each op you want to fold

```
class ShapeCastOpFolder : public RegionOpFolder<ShapeCastOpFolder> ...

class BroadcastOpFolder : public RegionOpFolder<ShapeCastOpFolder> ...
```

In these you can just check if any of the operands of this op is an argument to the block and use the `regionOperands` to get the replacement value to use. When you do that you end up using values from outside the region, so that breaks the `IsolatedFromAbove` property of `scf.for`, we will fix that later.

Add the pattern above to a pattern list `patterns` and run them on the region of the for op.

```
applyPatternsAndFoldGreedily(forOp->getRegion(), patterns)
```

After the patterns are applied you then see which values are used within the region of the `forOp` and use `getUsedValuesDefinedFromAbove` to get the values defined outside the `forOp` but used in the region. These are the "new" arguments of the `forOp`. Update the operands, and the arguments of the `forOp` region to fix up everything.

Essentially with this logic, you can extend the "folding" to as many ops as you want and never have to update the actual logic of how everything is worked out.  This might not pass muster in MLIR core since we break the isolated from above aspect for a bit, but it should work. THere are places in IREE we had to do this :)

Reply:
Example from Thomas 

```
%lr = for %a0 = %init
   cast %a0
....
   %r = cast %s
    yield %r
```
->
```
%init1 = cast %init 
%lr1 = for %a0 = %init1
    yield %s
%lr = cast %lr1
```

Reply:
Lets take a pattern that works on `shape_cast` above.

```
class ShapeCastOpFolder : public RegionOpFolder {
  LogicalResult matchAndRewrite(ShapeCastOp op, PatternRewriter &rewriter) {
    auto arg = op.source().dyn_cast<BlockArgument>();
    auto argOp = (arg ?
         dyn_cast<ShapeCastOp>(regionArguments[arg.getArgNumber()].getDefiningOp<ShapeCastOp>()
         : op.source().getDefiningOp<ShapeCastOp>();
    if (!argOp) return failure();
    if (argOp.source().getType() != op.getResult().getType()) return failure();
    rewriter.updateRootInPlace(op, [](Operation * op) { replaceAllUsesWith(argOp.source()); })
    return success();
  }
};
...
   {
  scf::ForOp forOp = ....
  patterns.insert<ShapeCastOpFolder>(context, forOp.getOperands());
  applyPatternsAndFoldGreedily(context, patterns);

  makeIsolatedFromAbove(forOp);
  fixReturnType(forOp);
}

```



With this you will end up with 

```
%21 = vector.shape_cast %20 : vector<4xf32> to vector<1x4xf32>
%22 = scf.for %arg3 = %c0 to %4097 step %c4 iter_args(%arg4 = %21) -> vector<1x4xf32> {
   ... %20 ...
  [...]
  scf.yield %108 : vector<4xf32>
}
```

This is wrong in two ways
1) The value `%20` is used without being passed as an argument
2) THe yield value and return type done match up.

So after you apply the patterns you have to fix up the `forOp` by
1) Changing the operands to be the values captured from above  (`makeIsolatedFromAbove`)
2) Changing the return type of the forOp. (`fixReturnType`)